### PR TITLE
Bug 1874457: Ensure router is present when Kuryr is used

### PR DESCRIPTION
**Pull request by MaysaMacedo**
Kuryr requires a router to be present on the cluster
to allow communication between Pod and Services. However,
during a FIP less no router is created by the installer.
This commit makes sure a router is created if not already
present.

### CODE REVIEW

**Comment by MaysaMacedo** [comment]
/hold

Still needs to be tested.

File: pkg/platform/openstack/networking.go [hunk 1]
```diff
@@ -183,7 +183,15 @@ func findOpenStackRouter(client *gophercloud.ServiceClient, name, tag string) (r
 	if len(routerList) == 1 {
 		return routerList[0], nil
 	} else if len(routerList) == 0 {
-		return empty, errors.New("router not found")
+		router, err := routers.Create(client, routers.CreateOpts{Name: name}).Extract()
```
Comment (luis5tb) [inline pkg/platform/openstack/networking.go:186]: we also need to connect the router to the VMs/Nodes network, right?

Comment (MaysaMacedo) [reply pkg/platform/openstack/networking.go:186]: True!

File: pkg/platform/openstack/networking.go [hunk 2]
```diff
@@ -167,9 +208,10 @@ func ensureOpenStackSubnet(client *gophercloud.ServiceClient, name, tag, netId,
 	}
 }
 
-// Looks for a Neutron router by name and tag. Fails if router is not found
-// or multiple routers match.
-func findOpenStackRouter(client *gophercloud.ServiceClient, name, tag string) (routers.Router, error) {
+// Looks for a Neutron router by name and tag. If not found, provides
+// the router used by the costum Network. If no router exists, creates
```
Comment (luis5tb) [inline pkg/platform/openstack/networking.go:212]: custom

File: pkg/platform/openstack/networking.go [hunk 3]
```diff
@@ -183,12 +225,53 @@ func findOpenStackRouter(client *gophercloud.ServiceClient, name, tag string) (r
 	if len(routerList) == 1 {
 		return routerList[0], nil
 	} else if len(routerList) == 0 {
-		return empty, errors.New("router not found")
+		ps, err := getOpenStackPortsByNetwork(client, networkID)
```
Comment (luis5tb) [inline pkg/platform/openstack/networking.go:228]: ps? perhaps worth to rename it to networkPorts?

File: pkg/platform/openstack/networking.go [hunk 4]
```diff
@@ -183,12 +225,53 @@ func findOpenStackRouter(client *gophercloud.ServiceClient, name, tag string) (r
...
+				if err != nil {
+					return empty, errors.Wrap(err, "failed to get router list")
+				}
+				routerList, err := routers.ExtractRouters(page)
+				if err != nil {
+					return empty, errors.Wrap(err, "failed to extract routers list")
+				}
+				if len(routerList) == 1 {
+					return routerList[0], nil
+				}
```
Comment (dulek) [inline pkg/platform/openstack/networking.go:244]: If we have the ID of the router we can just do `routers.Get` and simplify this.

**Review by dulek — CHANGES_REQUESTED** [review]
One small improvement that can be made.

File: pkg/platform/openstack/networking.go (hunk 4 above)
Comment (MaysaMacedo) [reply pkg/platform/openstack/networking.go]: Indeed. Thank you!

**Comment by MaysaMacedo** [comment]
Waiting for this [PR](https://github.com/openshift/installer/pull/4159) to get merged, before moving the hold.

**Pull request merged**